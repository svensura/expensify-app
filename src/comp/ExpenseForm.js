import React from 'react';
import moment from 'moment';
import 'react-dates/initialize';
import { SingleDatePicker } from 'react-dates';

export default class ExpenseForm extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            description: !!props.expense ? props.expense.description : '',
            note: !!props.expense ? props.expense.note : '',
            amount: !!props.expense ? (props.expense.amount / 100).toString() : '',
            createdAt: !!props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused: false,
            error: ''
        };
        this.onAmountChange = this.onAmountChange.bind(this);
        this.onDateChange = this.onDateChange.bind(this);
        this.onDescriptionChange = this.onDescriptionChange.bind(this);
        this.onFocusChange = this.onFocusChange.bind(this);
        this.onNoteChange = this.onNoteChange.bind(this);
        this.onSubmit = this.onSubmit.bind(this);
    }
    onAmountChange(e) {
        const amount = e.target.value;

        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({
                amount
            }));
        };
    }
    onDateChange(createdAt) {
        if (createdAt) {
            this.setState(() => ({
                createdAt
            }));
        };
    }
    onDescriptionChange(e) {
        const description = e.target.value;

        this.setState(() => ({
            description
        }));
    }
    onFocusChange({ focused }) {
        this.setState(() => ({
            calendarFocused: focused
        }));
    }
    onNoteChange(e) {
        const note = e.target.value;

        this.setState(() => ({
            note
        }));
    }
    onSubmit(e) {
        e.preventDefault();

        if (!this.state.description || !this.state.amount) {
            this.setState(() => ({
                error: 'Error.'
            }));
        } else {
            this.setState(() => ({
                error: ''
            }));

            this.props.onSubmit({
                description: this.state.description,
                note: this.state.note,
                amount: parseFloat(this.state.amount, 10) * 100,
                createdAt: this.state.createdAt.valueOf()
            });
        };
    }
    render() {
        return (
            <div>
                {!!this.state.error && <p>{this.state.error}</p>}
                <form onSubmit={this.onSubmit}>
                    <input
                        onChange={this.onDescriptionChange}
                        type="text"
                        placeholder="Description"
                        value={this.state.description}
                        autoFocus                    
                    />
                    <input
                        onChange={this.onAmountChange}
                        type="text"
                        placeholder="Amount"
                        value={this.state.amount}
                    />
                    <SingleDatePicker
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={() => false}
                    />
                    <textarea
                        onChange={this.onNoteChange}
                        placeholder="Describe Your Expense In Detail Here"
                        value={this.state.note}
                    >
                    </textarea>
                    <button>{!!this.props.expense ? 'Edit Expense' : 'Add Expense'}</button>
                </form>
            </div>
        );
    }
};