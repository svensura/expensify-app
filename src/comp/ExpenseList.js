import React from 'react';
import { connect } from 'react-redux';

import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

// pre-refactoring

/*
const ExpenseList = props => (
    <div>
        <h1>Expense List</h1>
        {
            props.expenses.map((expense, index) => {
                return (
                    <ExpenseListItem
                        key={index}
                        description={expense.description}
                        amount={expense.amount}
                        createdAt={expense.createdAt}
                    />
                )
            })
        }
    </div>
);
*/

// refactoring step II

export const ExpenseList = props => (
    <div>
        {
            props.expenses.length === 0 ? (
                <p>No expenses.</p>
            ) : (
                props.expenses.map((expense, index) => {
                    return (
                        <ExpenseListItem
                            key={expense.id}
                            {...expense}
                        />
                    )
                })              
            )
        }

    </div>
);

// pre-refactoring

/*
const ConnectedExpenseList = connect(state => {
    return {
        expenses: state.expenses
    };
})(ExpenseList);

export default ConnectedExpenseList;
*/

// refactoring step I

/*
export default connect(state => {
    return {
        expenses: state.expenses
    };
})(ExpenseList);
*/

// refactoring step II

const mapStateToProps = state => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    };
};

export default connect(mapStateToProps)(ExpenseList);