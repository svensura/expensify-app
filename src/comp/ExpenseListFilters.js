import React from 'react';

import { connect } from 'react-redux';
import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate } from '../actions/filters';

import 'react-dates/initialize';
import { DateRangePicker } from 'react-dates';

export class ExpenseListFilters extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            calendarFocused: null
        };
        this.onDatesChange = this.onDatesChange.bind(this);
        this.onFocusChange = this.onFocusChange.bind(this);
        this.onSortChange = this.onSortChange.bind(this);
        this.onTextChange = this.onTextChange.bind(this);
    }
    onDatesChange({ startDate, endDate }) {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    }
    onFocusChange(calendarFocused) {
        this.setState(() => ({
            calendarFocused
        }));
    }
    onSortChange(e) {
        e.target.value === 'date'
            ? this.props.sortByDate()
            : this.props.sortByAmount()
    }
    onTextChange(e) {
        this.props.setTextFilter(e.target.value);
    }
    render() {
        return (
            <div>
                <input
                    onChange={this.onTextChange}
                    type="text"
                    value={this.props.filters.text}
                />
                <select
                    onChange={this.onSortChange}
                    value={this.props.filters.sortBy}
                >
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>
                <DateRangePicker
                    onDatesChange={this.onDatesChange}
                    onFocusChange={this.onFocusChange}
                    startDate={this.props.filters.startDate}
                    startDateId={!!this.props.filters.startDate ? this.props.filters.startDate.toString() : 'none'}
                    endDate={this.props.filters.endDate}
                    endDateId={!!this.props.filters.endDate ? this.props.filters.endDate.toString() : 'none'}
                    focusedInput={this.state.calendarFocused}
                    showClearDates={true}
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                />
            </div>
        );
    }
};

const mapStateToProps = state => ({
    filters: state.filters
});

const mapDispatchToProps = dispatch => ({
    setTextFilter: text => dispatch(setTextFilter(text)),
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: () => dispatch(sortByAmount()),
    setStartDate: startDate => dispatch(setStartDate(startDate)),
    setEndDate: endDate => dispatch(setEndDate(endDate))
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);