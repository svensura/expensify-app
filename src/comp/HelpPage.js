import React from 'react';

const HelpPage = props => (
    <p>This. Is. Help. Page.</p>
);

export default HelpPage;