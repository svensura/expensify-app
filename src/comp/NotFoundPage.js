import React from 'react';
import { Link } from 'react-router-dom';

const NotFoundPage = props => (
    <div>
        <p>This. Is. 404.</p>
        <Link to="/">Home</Link>
    </div>
);

export default NotFoundPage;