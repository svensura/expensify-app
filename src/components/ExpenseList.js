import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses'


export const ExpenseList = (props) => (   // export to test the unconnected method
    <div className="content-container">
        <div className="list-header">
            <div className="show-for-mobile">Ausgaben</div>
            <div className="show-for-desktop">Ausgabe</div>
            <div className="show-for-desktop">Betrag</div>
        </div>
        <div className="list-body">
        {
            props.expenses.length === 0 ? (
                <div className="list-item--message"><span>Keine Ausgaben</span></div>
            ) : (
                props.expenses.map((expense) => (
                    <ExpenseListItem
                        key = {expense.id}
                        id = {expense.id}
                        description={expense.description} 
                        amount={expense.amount} 
                        createdAt ={expense.createdAt}
                    />
                    )
                )
            

            )
        }
        </div>
    </div>
)

const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    }
}

export default connect(mapStateToProps)(ExpenseList)
