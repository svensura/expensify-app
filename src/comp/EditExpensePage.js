import React from 'react';
import { connect } from 'react-redux';

import ExpenseForm from './ExpenseForm';
import { editExpense, removeExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component {
    constructor(props) {
        super(props);
        this.onSubmit = this.onSubmit.bind(this);
        this.onClick = this.onClick.bind(this);
    }
    onClick() {
        this.props.removeExpense({ id: this.props.expense.id })
        this.props.history.push('/');
    }
    onSubmit(expense) {
        this.props.editExpense(this.props.expense.id, expense)
        this.props.history.push('/');
    }
    render() {
        return (
            <div>
                <ExpenseForm
                    onSubmit={this.onSubmit}
                    expense={this.props.expense}
                />
                <button
                    onClick={this.onClick}
                >
                    Remove Expense
                </button>
            </div>
        );
    }
};

/*
const EditExpensePage = props => {
    return (
        <div>
            <ExpenseForm
                onSubmit={expense => {
                    console.log('updated', expense);
                    props.dispatch(editExpense(props.expense.id, expense));
                    props.history.push('/');
                }}
                expense={props.expense}
            />
            <button
                onClick={() => {
                    props.dispatch(removeExpense({ id: props.expense.id }));
                    props.history.push('/');
                }}
            >
                Remove Expense
            </button>
        </div>
    );
};
*/

const mapStateToProps = (state, props) => ({
    expense: state.expenses.find(expense => expense.id === props.match.params.id)
});

const mapDispatchToProps = (dispatch, props) => ({
    editExpense: (id, expense) => dispatch(editExpense(id, expense)),
    removeExpense: arg => dispatch(removeExpense(arg))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);