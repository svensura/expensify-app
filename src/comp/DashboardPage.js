import React from 'react';

import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';

const DashboardPage = props => (
    <div>
        <ExpenseListFilters />
        <ExpenseList />
    </div>
);

export default DashboardPage;