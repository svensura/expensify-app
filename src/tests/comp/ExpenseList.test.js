import React from 'react';
import { shallow } from 'enzyme';

import { ExpenseList } from '../../comp/ExpenseList';
import expenses from '../fixtures/expenses';

test('SHOULD RENDER <ExpenseList /> WITH expenses', () => {
    const wrapper = shallow(<ExpenseList expenses={expenses} />);
    expect(wrapper).toMatchSnapshot();
});

test('SHOULD RENDER <ExpenseList /> WITH EMPTY MESSAGE', () => {
    const wrapper = shallow(<ExpenseList expenses={[]} />);
    expect(wrapper).toMatchSnapshot();
});