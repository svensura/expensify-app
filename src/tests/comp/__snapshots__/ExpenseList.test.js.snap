// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SHOULD RENDER <ExpenseList /> WITH EMPTY MESSAGE 1`] = `
<div>
  <p>
    No expenses.
  </p>
</div>
`;

exports[`SHOULD RENDER <ExpenseList /> WITH expenses 1`] = `
<div>
  <ExpenseListItem
    amount={6000}
    createdAt={0}
    description="MacBook Pro 13"
    id="1"
    key="1"
    note="A new MacBook Pro 13"
  />
  <ExpenseListItem
    amount={5000}
    createdAt={5097600000}
    description="iPhone XS"
    id="2"
    key="2"
    note="A new iPhoneXS"
  />
  <ExpenseListItem
    amount={4000}
    createdAt={-10540800000}
    description="Xiaomi Mi Air 13"
    id="3"
    key="3"
    note="A new Xiaomi Mi Air 13"
  />
</div>
`;
